hail/src/main/scala/is/hail/expr/ir/MatrixIR.scala

```diff
@@ -164,97 +169,66 @@ case class MatrixNativeReader(path: String) extends MatrixReader {
+        TableNativeReader(colsPath, spec.colsTableSpec(colsPath))
+      )
 
-    val colAnnotations =
-      if (mr.dropCols)
-        FastIndexedSeq.empty[Annotation]
-      else
-        spec.colsComponent.readLocal(hc, path, requestedType.colType.physicalType).asInstanceOf[IndexedSeq[Annotation]]
+      tr = TableMapGlobals(tr, InsertFields(
+        Ref("global", tr.typ.globalType),
+        FastSeq(LowerMatrixIR.colsFieldName -> GetField(TableCollect(colsTable), "rows"))
+      ))
 
-    val rvd =
-      if (mr.dropRows)
-        RVD.empty(hc.sc, requestedType.canonicalRVDType)
-      else {
-        val fullRowType = requestedType.rvRowType
-        val localEntriesIndex = requestedType.entriesIdx
-
-        val rowsRVD = spec.rowsComponent.read(hc, path, requestedType.rowType.physicalType)
-        if (mr.dropCols) {
-          val (t2, makeF) = ir.Compile[Long, Long](
-            "row", requestedType.rowType.physicalType,
-            MakeStruct(
-              fullRowType.fields.zipWithIndex.map { case (f, i) =>
-                val v: IR = if (i == localEntriesIndex)
-                  MakeArray(FastSeq.empty, TArray(requestedType.entryType))
-                else
-                  GetField(Ref("row", requestedType.rowType), f.name)
-                f.name -> v
-              }))
-          assert(t2.virtualType == fullRowType)
-
-          rowsRVD.mapPartitionsWithIndex(requestedType.canonicalRVDType) { (i, it) =>
-            val f = makeF(i)
-            val rv2 = RegionValue()
-            it.map { rv =>
-              val off = f(rv.region, rv.offset, false)
-              rv2.set(rv.region, off)
-              rv2
-            }
-          }
-        } else {
-          val entriesRVD = spec.entriesComponent.read(hc, path, requestedType.entriesRVType.physicalType)
-          val entriesRowType = entriesRVD.rowType
-
-          val (t2, makeF) = ir.Compile[Long, Long, Long](
-            "row", requestedType.rowType.physicalType,
-            "entriesRow", entriesRowType.physicalType,
-            MakeStruct(
-              fullRowType.fields.zipWithIndex.map { case (f, i) =>
-                val v: IR = if (i == localEntriesIndex)
-                  GetField(Ref("entriesRow", entriesRowType), MatrixType.entriesIdentifier)
-                else
-                  GetField(Ref("row", requestedType.rowType), f.name)
-                f.name -> v
-              }))
-          assert(t2.virtualType == fullRowType)
-
-          rowsRVD.zipPartitionsWithIndex(requestedType.canonicalRVDType, entriesRVD) { (i, ctx, it1, it2) =>
-            val f = makeF(i)
-            val region = ctx.region
-            val rv3 = RegionValue(region)
-            new Iterator[RegionValue] {
-              def hasNext = {
-                val hn1 = it1.hasNext
-                val hn2 = it2.hasNext
-                assert(hn1 == hn2)
-                hn1
-              }
+      val entries: TableIR = TableRead(
+        TableType(
```

Comment:
Can't these types get pulled directly from the spec, too? Should we make sure they match?

Reply:
the isSupertype check should be sufficient